core/src/apps/cardano/helpers/tx_builder_state_machine.py

```diff
@@ -0,0 +1,144 @@
+            ACTION_ADD_OUTPUT_WITH_TOKENS: STATE_OUTPUT_WITH_TOKENS,
+            ACTION_FINISH_OUTPUTS: STATE_OUTPUTS_DONE,
+        },
+        STATE_OUTPUT_WITH_TOKENS: {
+            ACTION_START_ASSET_GROUPS: STATE_ASSET_GROUPS,
+        },
+        STATE_ASSET_GROUPS: {
+            ACTION_ADD_ASSET_GROUP: STATE_TOKENS,
+            ACTION_FINISH_ASSET_GROUPS: STATE_ASSET_GROUPS_DONE,
+        },
+        STATE_TOKENS: {
+            ACTION_ADD_TOKEN: STATE_TOKENS,
+            ACTION_FINISH_TOKENS: STATE_ASSET_GROUPS,
+        },
+        STATE_ASSET_GROUPS_DONE: {ACTION_FINISH_OUTPUT: STATE_OUTPUTS},
+        STATE_OUTPUTS_DONE: {ACTION_ADD_FEE: STATE_FEE_DONE},
+        STATE_FEE_DONE: {
+            ACTION_ADD_TTL: STATE_TTL_DONE,
+            ACTION_START_CERTIFICATES: STATE_CERTIFICATES,
+            ACTION_START_WITHDRAWALS: STATE_WITHDRAWALS,
+            ACTION_ADD_AUXILIARY_DATA_HASH: STATE_AUXILIARY_DATA_DONE,
+            ACTION_ADD_VALIDITY_INTERVAL_START: STATE_END,
+        },
+        STATE_TTL_DONE: {
+            ACTION_START_CERTIFICATES: STATE_CERTIFICATES,
+            ACTION_START_WITHDRAWALS: STATE_WITHDRAWALS,
+            ACTION_ADD_AUXILIARY_DATA_HASH: STATE_AUXILIARY_DATA_DONE,
+            ACTION_ADD_VALIDITY_INTERVAL_START: STATE_END,
+        },
+        STATE_CERTIFICATES: {
+            ACTION_ADD_SIMPLE_CERTIFICATE: STATE_CERTIFICATES,
+            ACTION_ADD_POOL_REGISTRATION_CERTIFICATE: STATE_POOL_REGISTRATION_CERTIFICATE,
+            ACTION_FINISH_CERTIFICATES: STATE_CERTIFICATES_DONE,
+        },
+        STATE_POOL_REGISTRATION_CERTIFICATE: {
+            ACTION_START_POOL_OWNERS: STATE_POOL_OWNERS
+        },
+        STATE_POOL_OWNERS: {
+            ACTION_ADD_POOL_OWNER: STATE_POOL_OWNERS,
+            ACTION_FINISH_POOL_OWNERS: STATE_POOL_OWNERS_DONE,
+        },
+        STATE_POOL_OWNERS_DONE: {ACTION_START_POOL_RELAYS: STATE_POOL_RELAYS},
+        STATE_POOL_RELAYS: {
+            ACTION_ADD_POOL_RELAY: STATE_POOL_RELAYS,
+            ACTION_FINISH_POOL_RElAYS: STATE_POOL_RELAYS_DONE,
+        },
+        STATE_POOL_RELAYS_DONE: {ACTION_FINISH_CERTIFICATE: STATE_CERTIFICATES},
+        STATE_CERTIFICATES_DONE: {
+            ACTION_START_WITHDRAWALS: STATE_WITHDRAWALS,
+            ACTION_ADD_AUXILIARY_DATA_HASH: STATE_AUXILIARY_DATA_DONE,
+            ACTION_ADD_VALIDITY_INTERVAL_START: STATE_END,
+        },
+        STATE_WITHDRAWALS: {
+            ACTION_ADD_WITHDRAWAL: STATE_WITHDRAWALS,
+            ACTION_FINISH_WITHDRAWALS: STATE_WITHDRAWALS_DONE,
+        },
+        STATE_WITHDRAWALS_DONE: {
+            ACTION_ADD_AUXILIARY_DATA_HASH: STATE_AUXILIARY_DATA_DONE,
+            ACTION_ADD_VALIDITY_INTERVAL_START: STATE_END,
+        },
+        STATE_AUXILIARY_DATA_DONE: {ACTION_ADD_VALIDITY_INTERVAL_START: STATE_END},
+    }
+
+    def __init__(self) -> None:
+        self.state = self.STATE_INIT
+
+    def transition(self, action: ActionType) -> None:
+        assert (
```

Comment:
Sending the wrong element would fail in the given `sign_tx._process_xxx` function when calling `await ctx.call(CardanoTxItemAck(), XXXExpectedMessage)` because the `XXXExpectedMessage` would not be received. Thus if the assertion is false it's likely due to programmer error.